File: 0542-01-matrix/0542-01-matrix.ts
function updateMatrix(mat: number[][]): number[][] {
    const rows=mat.length;
    const cols=mat[0].length;
    const queue:[number,number][]=[];
    const dist:number[][]=Array.from(
        {length:rows},
        ()=>Array(cols).fill(-1)
    );

    //Initialize queue with all 0s
    for(let r=0;r<rows;r++){
        for(let c=0;c<cols;c++){
            if(mat[r][c]===0){
                queue.push([r,c]);
                dist[r][c]=0;
            }
        }
    }
    const directions=[
        [1,0],
        [-1,0],
        [0,1],
        [0,-1]    ];

    let front=0;
    while(front<queue.length){
        const [r,c]=queue[front++];
        for(const [dr,dc] of directions){
            const nr=r+dr;
            const nc=c+dc;
            if(nr>=0&&nr<rows&&nc>=0&&nc<cols&&dist[nr][nc]===-1){
                dist[nr][nc]=dist[r][c]+1;
                queue.push([nr,nc]);
            }
        }
    }

return dist;


};

File: 0542-01-matrix/0542-01-matrix.java
import java.util.*;

class Solution {
    public int[][] updateMatrix(int[][] mat) {
        int m=mat.length;
        int n=mat[0].length;
        Queue<int[]> queue=new LinkedList<>();
        int[][] dist=new int[m][n];
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(mat[i][j]==0){
                    queue.offer(new int[]{i,j});
                }
                else{
                    dist[i][j]=-1;
                }
            }
        }
        int[][] directions={
            {1,0},{-1,0},{0,1},{0,-1}
        };

        while(!queue.isEmpty()){
            int[] curr=queue.poll();
            int row=curr[0];
            int col=curr[1];
            for(int[]dir:directions){
                int newRow=row+dir[0];
                int newCol=col+dir[1];
                if(newRow>=0 && newRow<m&&newCol>=0 && newCol<n&&dist[newRow][newCol]==-1){
                    dist[newRow][newCol]=dist[row][col]+1;
                    queue.offer(new int[]{newRow,newCol});

                }
            }


        }


        return dist;



    }
}